src/kepler_model/train/profiler/node_type_index.py

```diff
@@ -183,13 +233,34 @@ def index_train_machine(self, machine_id, new_spec):
         self.node_type_index[covered_index].add_member(machine_id)
         return covered_index
 
-    def get_node_type(self, compare_spec):
+    def get_node_type(self, in_spec, loose_search=False, debug=False):
```

Comment:
this function is a bit too complex. Is it possible for us
* to break it down into smaller units
* write a unit-test ?